Azure Core Release Beta 7

azure-sdk opened the pull request:
Update vcpkg ports for Azure SDK release. This release may contain multiple ports.

azure-sdk committed:
[azure-core-cpp] Update to 1.0.0-beta.7
## 1.0.0-beta.7 (2021-03-11)

### New Features

- Added `HttpPolicyOrder` for adding custom Http policies to SDK clients.
- Added `Azure::Core::Operation<T>::GetRawResponse()`.
- Added `Azure::Core::PackageVersion`.
- Added support for logging to console when `AZURE_LOG_LEVEL` environment variable is set.

### Breaking Changes

- Changes to `Azure::Core` namespace:
  - Removed Response<void>, `ValueBase`, and `ContextValue`.
  - Removed `Context::operator[]`, `Get()` introduced instead.
  - Renamed `Uuid::GetUuidString()` to `ToString()`.
  - Changed return type of `Operation<T>::Poll()` from `std::unique_ptr<RawResponse>` to `RawResponse const&`.
  - Moved `GetApplicationContext()` to `Context::GetApplicationContext()`
  - Moved the `Base64Encode()` and `Base64Decode()` functions to be static members of a `Convert` class.
  - Moved `Logging` namespace entities to `Diagnostics::Logger` class.
  - Moved `AccessToken`, `TokenCredential`, and `AuthenticationException` to `Credentials` namespace.
  - Moved `Context` to be the last parameter for consistency, instead of first in various azure-core types. For example:
    - `BodyStream::Read(uint8_t* buffer, int64_t count, Context const& context)`
    - `BodyStream::ReadToEnd(BodyStream& body, Context const& context)`
    - `HttpPolicy::Send(Request& request, NextHttpPolicy policy, Context const& context)`
    - `Operation<T>::PollUntilDone(std::chrono::milliseconds period, Context& context)`
    - `TokenCredential::GetToken(Http::TokenRequestOptions const& tokenRequestOptions, Context const& context)`
  - Moved from `Azure::Core` to `Azure` namespace:
    - `Response<T>`, `ETag`, and `Nullable<T>`.
    - Split `RequestConditions` into `MatchConditions` and `ModifiedConditions`.
    - Renamed `DateTime::GetString()` to `ToString()`, and removed `DateTime::GetRfc3339String()`.
- Changes to `Azure::Core::Http` namespace:
  - Removed `HttpPipeline`, `TransportKind`, `NullBodyStream`, and `LimitBodyStream`.
  - Removed `Request::StartTry()`.
  - Removed `InvalidHeaderException` and throw `std::invalid_argument` if the user provides invalid header arguments.
  - Renamed `CurlTransportSSLOptions::NoRevoke` to `EnableCertificateRevocationListCheck`.
  - Renamed `Range` to `HttpRange`.
  - Renamed `TokenRequestOptions` to `TokenRequestContext`, and moved it to `Azure::Core::Credentials` namespace.
  - Moved `Url` to `Azure::Core` namespace.
  - `Request` and `RawResponse`:
    - Renamed `AddHeader()` to `SetHeader()`.
    - Introduced `Azure::Core::CaseInsensitiveMap` which is now used to store headers.
  - `BodyStream` and the types that derive from it:
    - Moved to `Azure::Core::IO` namespace.
    - Changed the static methods `BodyStream::ReadToCount()` and `BodyStream::ReadToEnd()` into instance methods.
    - Changed the constructor of `FileBodyStream` to accept a file name directly and take ownership of opening/closing the file, instead of accepting a file descriptor, offset, and length.
  - HTTP policies and their options:
    - Moved to `Policies` namespace.
    - Renamed `TransportPolicyOptions` to `TransportOptions`.
    - Renamed `TelemetryPolicyOptions` to `TelemetryOptions`.
    - Changed type of `RetryOptions::StatusCodes` from `std::vector` to `std::set`.
    - Renamed `LoggingPolicy` to `LogPolicy`, and introduced `LogOptions` as mandatory parameter for the constructor.
- Moved header files:
  - Renamed `azure/core/credentials.hpp` to `azure/core/credentials/credentials.hpp`.
  - Renamed `azure/core/logger.hpp` to `azure/core/diagnostics/logger.hpp`.
  - Renamed `azure/core/http/body_stream.hpp` to `azure/core/io/body_stream.hpp`.
  - Renamed `azure/core/http/policy.hpp` to `azure/core/http/policies/policy.hpp`.
  - Renamed `azure/core/http/curl/curl.hpp` to `azure/core/http/curl_transport.hpp`.
  - Renamed `azure/core/http/winhttp/win_http_client.hpp` to `azure/core/http/win_http_transport.hpp`.

### Bug Fixes

- Make sure to rewind the body stream at the start of each request retry attempt, including the first.
- Connection pool resets when all connections are closed.
- Fix `Azure::Context` to support `std::unique_ptr`.
- Throw `std::runtime_error` from `Response<T>::GetRawResponse()` if the response was already extracted.

azure-sdk committed:
[azure-identity-cpp] Update to 1.0.0-beta.4
## 1.0.0-beta.4 (2021-03-11)

### New Features

- Added `Azure::Identity::PackageVersion`.

### Breaking Changes

- Removed `TransportPolicyOptions` from `ClientSecretCredentialOptions`. Updated the options to derive from ClientOptions.

ahsonkhan commented:
Given the breaking changes in core for this beta release of azure-core, upstream package dependencies like azure-storage-common are expected to fail.

We either need to ship the storage package releases to this same PR/branch, OR break the dependency cycle and update the .\scripts\ci.baseline.txt file to expect storage packages to be broken for now (until the newer versions of those can be published), in order to unblock this PR and rev azure-core and identity.

cc @antkmsft, @RickWinter, @Jinming-Hu, @danieljurek, @vhvb1989

ahsonkhan committed:
Skip building azure-storage-*-cpp packages to break dependency cycle and
publish azure-core-cpp.

NancyLi1013 commented on scripts/ci.baseline.txt:
```diff
@@ -50,6 +50,44 @@ apr:arm64-windows=fail
+azure-storage-files-datalake-cpp:x64-windows-static=skip
+azure-storage-files-datalake-cpp:x64-windows-static-md=skip
+azure-storage-files-datalake-cpp:x86-windows=skip
+
```
So the new update for `azure-core-cpp` will block these ports on the triplets, right?

ahsonkhan replied:
Yes. We will be releasing updates to these ports soon, from https://github.com/Azure/azure-sdk-for-cpp.

ras0219-msft commented:
We would strongly prefer:
1. Merging the new versions of the downstream dependents first (assuming they will be compatible with the current version of azure-identity et al).
2. All the updates being in the same PR, enabling an "atomic" update of the related ports.

We are strongly against merging breaking changes before fixes, because it means that customers which fetch the intermediate commit will be broken and unable to install `azure-storage-files-datalake-cpp`.

ahsonkhan commented:
> Merging the new versions of the downstream dependents first (assuming they will be compatible with the current version of azure-identity et al).

Since the ports are currently in beta, the downstream dependents won't work against the current versions.

> All the updates being in the same PR, enabling an "atomic" update of the related ports.

That makes sense. We'll try to follow this approach.

azure-sdk committed:
[azure-storage-common-cpp] Update to 12.0.0-beta.9
## 12.0.0-beta.9 (2021-03-23)

### New Features

- Added `Azure::Storage::Common::PackageVersion`.

azure-sdk committed:
[azure-storage-blobs-cpp] Update to 12.0.0-beta.9
## 12.0.0-beta.9 (2021-03-23)

### New Features

- Added support for telemetry options.
- Added `Azure::Storage::Blobs::PackageVersion`.
- Added `ShareFileClient::GetRangeListDiff`.

### Breaking Changes

- Changed the return type of `StartCopyFromUri` and `StartCopyIncremental` API from a `Response<T>` to the particular `Operation<T>` type called `StartBlobCopyOperation` directly.
- String conversion functions of extensible enums were renamed from `Get()` to `ToString()`.
- Moved `SecondaryHostForRetryReads` out of retry options, now it's under `BlobClientOptions`.
- Changed the return types of the following APIs:
  - `BlobServiceClient::DeleteBlobContainer` now returns `DeleteBlobContainerResult`.
  - `BlobServiceClient::GetUserDelegationKey` now returns `UserDelegationKey`.
  - `BlobServiceClient::GetProperties` now returns `BlobServiceProperties`.
  - `BlobServiceClient::GetAccountInfo` now returns `AccountInfo`.
  - `BlobServiceClient::GetStatistics` now returns `ServiceStatistics`.
  - `BlobContainerClient::DeleteBlob` now returns `DeleteBlobResult`.
  - `BlobContainerClient::GetProperties` now returns `BlobContainerProperties`.
  - `BlobContainerClient::GetAccessPolicy` now returns `BlobContainerAccessPolicy`.
  - `BlobClient::GetProperties` now returns `BlobProperties`.
  - `BlobClinet::GetTags` now returns `std::map<std::string, std::string>`.
  - `BlobClient::AbortCopyFromUri` now returns `AbortBlobCopyFromUriResult`.
  - `PageBlobClient::UploadPages` now returns `UploadPagesResult`.
  - `PageBlobClient::UploadPagesFromUri` now returns `UploadPagesFromUriResult`.
  - `PageBlobClient::ClearPages` now returns `ClearPagesResult`.
  - `PageBlobClient::GetPageRanges`, `PageBlobClient::GetPageRangesDiff` and `PageBlobClient::GetManagedDiskPageRangesDiff` now return `GetPageRangesResult`.
- Renamed some options:
  - Renamed `UploadPageBlobPagesOptions` to `UploadPagesOptions`.
  - Renamed `UploadPageBlobPagesFromUriOptions` to `UploadPagesFromUriOptions`.
  - Renamed `ClearPageBlobPagesOptions` to `ClearPagesOptions`.
  - Renamed `GetPageBlobPageRangesOptions` to `GetPageRangesOptions`.
  - Renamed `StartCopyBlobFromUriOptions` to `StartBlobCopyFromUriOptions`.
  - Renamed `StartCopyPageBlobIncrementalOptions` to `StartBlobCopyIncrementalOptions`.
- Removed `Blob` from the name of lease options and return types, like `AcquireBlobLeaseOptions` was renamed to `AcquireLeaseOptions`.
- Removed `Blob` prefix from the name of structs `BlobGeoReplication`, `BlobLeaseStatus`, `BlobLeaseState`, `BlobLeaseDurationType`, `BlobAnalyticsLogging`, `BlobMetrics`, `BlobCorsRule`, `BlobRetentionPolicy`, `BlobStaticWebsite`, `BlobArchiveStatus` and `BlobGeoReplictionStatus`.
- Removed `RequestId` from return types.
- Changed `BodyStream` parameter of `UploadBlob`, `Upload`, `UploadPages` and `AppendBlock`functions from pointer to reference.
- Renamed access tier members in return types and options from `Tier` to `AccessTier`.
- Renamed `BlockListTypeOption` to `BlockListType`.

azure-sdk committed:
[azure-storage-files-datalake-cpp] Update to 12.0.0-beta.9
## 12.0.0-beta.9 (2021-03-23)

### New Features

- Added support for telemetry options.
- Added `Azure::Storage::Files::DataLake::PackageVersion`.

### Breaking Changes

- DataLake client constructors won't automatically convert blob url to dfs url anymore.
- String conversion functions of extensible enums were renamed from `Get()` to `ToString()`.
- Moved `SecondaryHostForRetryReads` out of retry options, now it's under `DataLakeClientOptions`.
- Changed the return types of the following APIs:
  - `DataLakeServiceClient::GetUserDelegationKey` now returns `UserDelegationKey`.
  - `DataLakeFileSystemClient::GetProperties` now returns  `FileSystemProperties`.
  - `DataLakeFileSystemClient::GetAccessPolicy` now returns  `FileSystemAccessPolicy`.
  - `DataLakePathClient::GetProperties` now returns `PathProperties`.
  - `DataLakePathClient::GetAccessControlList` now returns `PathAccessControlList`.
- Removed `GetUserDelegationKeyResult`.
- Removed `DataLake` from the names of return types and option types.
- Removed `RequestId` from the return types.
- Changed `BodyStream` parameter of `Append` function from pointer to reference.
- Removed `PathRenameMode`, `PathGetPropertiesAction`, `PathSetAccessControlRecursiveMode`, `FileSystemResourceType`, `PathExpiryOptions` and `FileSystemResourceType`.
- Removed `IsAccessTierInferred` and `AccessTierChangedOn` from `PathProperties`.
- Renamed `LeaseDurationType` to `LeaseDuration`, `LeaseStateType` to `LeaseState` and `LeaseStatusType` to `LeaseStatus`.

azure-sdk committed:
[azure-storage-files-shares-cpp] Update to 12.0.0-beta.9
## 12.0.0-beta.9 (2021-03-23)

### New Features

- Added support for telemetry options.
- Added `Azure::Storage::Files::Shares::PackageVersion`.

### Breaking Changes

- Changed the return type of `StartCopy` API from a `Response<T>` to the particular `Operation<T>` type called `StartFileCopyOperation` directly.
- String conversion functions of extensible enums were renamed from `Get()` to `ToString()`.
- Changed the return types of the following APIs:
  - `ShareClient::GetAccessPolicy` now returns `ShareAccessPolicy`.
  - `ShareClient::GetPermission` now returns `std::string`.
  - `ShareClient::AbortCopy` now returns `AbortFileCopyResult`.
- Renamed `GetShareStatisticsResult` to `ShareStatistics`.
- Renamed `GetSharePropertiesResult` to `ShareProperties`.
- Renamed `GetShareDirectoryPropertiesResult` to `DirectoryProperties`.
- Renamed `GetShareFilePropertiesResult` to `FileProperties`
- Renamed `GetServicePropertiesResult` to `ShareServiceProperties`.
- Removed `Share` from the names of return types and option types.
- Renamed `AbortCopyFileOptions` to `AbortFileCopyOptions`.
- Removed `RequestId` from the return types.
- Changed `BodyStream` parameter of `UploadRange` function from pointer to reference.
- Removed `PreviousShareSnapshot` from `GetShareFileRangeListOptions`, use `ShareFileClient::GetRangeListDiff` instead.
- Renamed `ShareAccessTier` to `AccessTier`.
- Renamed `ShareRetentionPolicy` to `RetentionPolicy`.
- Renamed `ShareProtocolSettings` to `ProtocolSettings`.
- Renamed `CopyStatusType` to `CopyStatus`
- Removed `FileRangeWriteType`, `ShareFileRangeList`, `FileRangeWriteFromUrlType`, `FileRange`, `ClearRange`, `SharePermission`, `LeaseAction` and `ShareStats`.
- Renamed `LeaseDurationType` to `LeaseDuration`, `LeaseStateType` to `LeaseState` and `LeaseStatusType` to `LeaseStatus`.
- Renamed `ListSharesIncludeType` to `ListSharesIncludeFlags`.
- Renamed `DeleteSnapshotsOptionType` to `DeleteSnapshotsOption`.
- Renamed `PermissionCopyModeType` to `PermissionCopyMode`.

ahsonkhan commented on scripts/ci.baseline.txt:
```diff
@@ -50,6 +50,44 @@ apr:arm64-windows=fail
 # Requires ATL for ARM64 to be installed in CI
 azure-storage-cpp:arm64-windows=fail
 
+# These depend on azure-core-cpp which had breaking changes and need to be rev'd to newer beta versions.
+azure-storage-common-cpp:arm64-windows=skip
+azure-storage-common-cpp:arm-uwp=skip
```
We can now revert this change from commit d9b75cf1348e1acdaef56f678bd3cd49af52aaa3

ahsonkhan committed:
Revert "Skip building azure-storage-*-cpp packages to break dependency cycle and"

This reverts commit d9b75cf1348e1acdaef56f678bd3cd49af52aaa3.

ahsonkhan committed:
Merge branch 'azure-sdk-for-cpp-march-2021' of https://github.com/azure-sdk/vcpkg into azure-sdk-for-cpp-march-2021

ahsonkhan commented:
@NancyLi1013 and @ras0219 please review this PR as I believe it is ready to be merged.

NancyLi1013 commented:
Hi @ahsonkhan 

Have you tested these features in this PR?

ahsonkhan commented:
> Have you tested these features in this PR?

Not directly, particularly because the release didn't modify any of the features or how they are exposed. We had tested the features in the previous release.

I would appreciate it if you could help test them this time around, and also please point me to the types of steps one would have to take to test the features in the future.

NancyLi1013 commented:
I checked these files again and found that only `azure-core-cpp` has features. Since `http ` has been set as default feature, which will depends on `curl `and `winhttp` features to build. Vcpkg will test default features on CI test. So in this case, it's unnecessary to test these features again.

Except for default features, other features need to test manually if your changes are related with these features, especially for bug fix or new version update. If the changes will not affect the functions of these features, you can make a note to describe this.

NancyLi1013 commented:
LGTM now, thanks for your updates @ahsonkhan.
